Classes/Domain/Repository/HaikuRepository.php

```diff
@@ -0,0 +1,60 @@
+<?php
+
+declare(strict_types=1);
+
+namespace T3docs\Examples\Domain\Repository;
+
+use T3docs\Examples\Exception\NoSuchHaikuException;
+use TYPO3\CMS\Core\Database\Connection;
+use TYPO3\CMS\Core\Database\ConnectionPool;
+
+class HaikuRepository
+{
+    private const TABLENAME = 'tx_examples_haiku';
+    private Connection $connection;
+
+    public function __construct(
+        ConnectionPool $connectionPool,
+    ) {
+        $this->connection = $connectionPool->getConnectionForTable(self::TABLENAME);
+    }
+
+    public function findAll(): array
+    {
+        $queryBuilder = $this->connection->createQueryBuilder();
+        $result = $queryBuilder->select('*')->from(self::TABLENAME)->executeQuery();
+        return $result->fetchAllAssociative();
+    }
+
+    public function findByUid(int $uid): array
+    {
+        $queryBuilder = $this->connection->createQueryBuilder();
+        // $uid is an integer so we don't have to worry about SQL injections
+        $where = $queryBuilder->expr()->eq('uid', $uid);
+        $result = $queryBuilder->select('*')->from(self::TABLENAME)->where(
+            $where
+        )->executeQuery()->fetchAssociative();
+        if (!$result) {
+            throw new NoSuchHaikuException('Haiku with uid ' . $uid . 'not found.');
```

Comment:
Thought about it again this day: Throwing this exception for an "empty" result (either `false` or `[]`) results in a 500 Server Error (problem is on server side). For `false` this is correct, as the query may be wrong from the developer side. But: For a not found uid (`[]`) there should be a 404. So we should handle this separate: `if (!is_array($result))` or (`if ($result === false)`) throws a custom exception (which results in a 500 status code) whereas an empty array should be returned to the controller and handled there (so a 404 status code can be returned via PropagateResponseException). I would prefer the `is_array` check, as an array must be returned.
This is relevant for SEO purposes, so we should give developers a good example in the docs.